staging/src/k8s.io/client-go/tools/leaderelection/healthzadaptor_test.go

```diff
@@ -0,0 +1,112 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package leaderelection
+
+import (
+	"testing"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/clock"
+	rl "k8s.io/client-go/tools/leaderelection/resourcelock"
+	"net/http"
+)
+
+type fakeLock struct {
+	identity string
+}
+
+// Get is a dummy to allow us to have a fakeLock for testing.
+func (fl *fakeLock) Get() (ler *rl.LeaderElectionRecord, err error) {
+	return nil, nil
+}
+
+// Create is a dummy to allow us to have a fakeLock for testing.
+func (fl *fakeLock) Create(ler rl.LeaderElectionRecord) error {
+	return nil
+}
+
+// Update is a dummy to allow us to have a fakeLock for testing.
+func (fl *fakeLock) Update(ler rl.LeaderElectionRecord) error {
+	return nil
+}
+
+// RecordEvent is a dummy to allow us to have a fakeLock for testing.
+func (fl *fakeLock) RecordEvent(string) {}
+
+// Identity is a dummy to allow us to have a fakeLock for testing.
+func (fl *fakeLock) Identity() string {
+	return fl.identity
+}
+
+// Describe is a dummy to allow us to have a fakeLock for testing.
+func (fl *fakeLock) Describe() string {
+	return "Dummy implementation of lock for testing"
+}
+
+// TestLeaderElectionHealthChecker tests that the healthcheck for leader election handles its edge cases.
+func TestLeaderElectionHealthChecker(t *testing.T) {
+	current := time.Now()
+	checker := NewLeaderHealthzAdaptor(time.Second * 20)
+	lc, ok := checker.(*healthzAdaptor)
+	if !ok {
+		t.Errorf("received checker of type %T not leaderChecker", checker)
+	}
+	if lc.le != nil {
+		t.Errorf("leaderChecker started with a LeaderElector %v", lc.le)
+	}
+	req := &http.Request{}
+	err := lc.Check(req)
+	if err != nil {
+		t.Errorf("received error checking status on uninitialized %v", err)
+	}
+	lec := LeaderElectionConfig{
+		Lock:          &fakeLock{identity: "healthTest"},
+		LeaseDuration: time.Minute,
+		WatchDog:      checker,
+		Name:          "foo",
+	}
+	// Setting the current time far enough in the future that the lease shows as expired.
+	le := &LeaderElector{
+		config: lec,
+		observedRecord: rl.LeaderElectionRecord{
+			HolderIdentity: "healthTest",
+		},
+		observedTime: current,
+		timer:        clock.NewFakeClock(current.Add(time.Hour)),
+	}
+	lc.SetLeaderElection(le)
+	err = lc.Check(req)
+	if err == nil {
+		t.Errorf("should have received error as leader election is enabled, we're leader and expired.")
+	}
+	le.observedRecord.HolderIdentity = "otherServer"
+	err = lc.Check(req)
+	if err != nil {
+		t.Errorf("received error checking status, we're not the leader %v", err)
+	}
+	le.observedRecord.HolderIdentity = "healthTest"
+	// Test for lease not yet expired
+	le.timer = clock.NewFakeClock(current)
```

Comment:
I'd use Step(-time.Hour) if you want to go backwards. Replacing the clock will break any uses of NewTimer, NewTicker, After, etc.

Actually if you just want to reset to set up the next scenario I'd probably rebuild the structures and/or make a separate test function, since time shouldn't go backwards unless this is a DST test or something :)

Reply:
Rebuilt